[DOCS] Add runtime fields to index templates (#70172)
elastic/elasticsearch: docs/reference/indices/index-templates.asciidoc

for `logs-*-*`.
====
// end::built-in-index-templates[]
****

[discrete]
[[create-index-templates]]
== Create composable index templates

> 

Reviewer: I think the focus needs to be on creating an index template. Since you want to highlight the Runtime field mapping, it might be easier to frame it as defining it directly in the index template, rather than in a component template for the UI piece.

> To create a component template from Stack Management in {kib}, select
> *Index Management* in the side navigation and choose *Component Templates*. You
> can then define the index settings, mappings, and alias configurations for the
> component template.

Reviewer: ```suggestion
To create an index template from Stack Management in {kib}, select
*Index Management* in the side navigation, choose *Index Templates*, and click *Create Template*. You
can then select the component templates you want to use, or define the index settings, mappings, and alias configurations directly.
```

When defining mappings, you can add mapped fields and create
<<runtime-mapping-fields,runtime fields>>. These fields work like any other
field, but are evaluated only at query time. When an index is created that
matches the index template, {es} uses the index template to configure the index
and add the runtime field to the mappings. When you create the runtime field,
you optionally specify a Painless script. If you don't include a script,
runtime fields retrieve values from `_source`.

> You can also use the <<indices-component-template,put component template API>>
> to create or update a component template. The following requests create two
> composable templates.

Reviewer: ```suggestion
You can also use the <<indices-put-template, index template>> and <<indices-component-template,put component template>> APIs to create and update index templates and template components.
The following requests create two
component templates.
```

  }
}

PUT _component_template/runtime_component_template
{
  "template": {
    "mappings": {
      "runtime": { <1>
        "day_of_week": {
          "type": "keyword",
          "script": {
            "source": "emit(doc['@timestamp'].value.dayOfWeekEnum.getDisplayName(TextStyle.FULL, Locale.ROOT))"
          }
        }
      },
      "properties": {

>         "@timestamp": {

Reviewer 2: because it's a component template, timestamp does not necessarily need to be defined in the same template?

Author: @javanna, can you please elaborate? I don't understand the connection between the timestamp and component template.

Reviewer 2: I was referring to the fact that the timestamp is already defined in the other component template, and you replaced the previous ip_address in this component template with the timestamp. I thought that is not necessary....but am I missing something?

Reviewer 3: It's also a very common pattern for users to have separate component templates for mappings and settings. For example, our built-in component templates do this: https://www.elastic.co/guide/en/elasticsearch/reference/master/index-management-settings.html#stack-templates-enabled

Author: No, no, you're correct. I was completely overlooking the timestamp in the other component template. I'll remove this one 👍

Reviewer 3: NVM. I misread. Please ignore my drive-by comment.

  }
}

PUT _component_template/runtime_component_template
{
  "template": {
    "mappings": {
      "runtime": { <1>
        "day_of_week": {
          "type": "keyword",
          "script": {
            "source": "emit(doc['@timestamp'].value.dayOfWeekEnum.getDisplayName(TextStyle.FULL, Locale.ROOT))"
          }
        }
      },
      "properties": {
        "@timestamp": {
          "type": "date"
        }
      }
    }
  }
}
----

> <1> This composable template adds a runtime field named `day_of_week` to the

Reviewer: ```suggestion
<1> This component template adds a runtime field named `day_of_week` to the
```

mappings when a new index matches the template.

> The following request creates an index template that contains the previous
> composable templates. Specify `"composed_of"` in the request to indicate the
> component templates that this index template contains.

Reviewer: ```suggestion
The following request creates an index template that is _composed of_  these
component templates. 
```
